src/rust/engine/process_execution/src/immutable_inputs.rs

```diff
@@ -2,33 +2,78 @@ use std::collections::{BTreeMap, HashMap};
 use std::path::{Path, PathBuf};
 use std::sync::Arc;
 
-use double_checked_cell_async::DoubleCheckedCell;
+use async_oncecell::OnceCell;
 use fs::{Permissions, RelativePath};
+use futures::TryFutureExt;
 use hashing::Digest;
 use parking_lot::Mutex;
 use store::Store;
 use tempfile::TempDir;
 
 use crate::WorkdirSymlink;
 
+async fn rename_readonly_directory(
+  src: impl AsRef<Path>,
+  dest: impl AsRef<Path>,
+  map_rename_err: impl Fn(std::io::Error) -> String,
+) -> Result<(), String> {
```

Comment:
If we end up needing to do this for any other materialized directories, we might be able to bake it into `materialize_directory` itself, before the root ends up marked readonly.

Reply:
Yeah, maybe. The materialize and move case that this use has would need a callback or an expclict parameter for a final dest.